os/board/rtl8730e/src/component/soc/amebad2/fwlib/ap_peripheral/ameba_loguart.c

```diff
@@ -34,16 +34,6 @@ void LOGUART_PutChar(u8 c)
 	u32 CounterIndex = 0;
 	u32 CPUID = SYS_CPUID();
 
-	while (1) {
-		CounterIndex++;
-		if (CounterIndex >= LOG_UART_IDX_FLAG[CPUID].tx_timeout) {
-			break;
-		}
-
-		if (UARTLOG->LOGUART_UART_LSR & LOG_UART_IDX_FLAG[CPUID].not_full) {
```

Comment:
Dear Mr @sunghan-chang, here is what i meant. it does not do not full check anymore, so it will require the check at up_putc

Reply:
Ok. You mean checking not full at up_putc, not here. But any reason? This uart hw is shared resource for all threads. And here is last point before writing. I don't know what will be better by moving the location. 
Could you elaborate more?

Reply:
Before this PR, rtl8730e_log_up_txreadywill always return 1, this assumes that fifo always have space for next data, the check is done at last point (loguart_putchar). however, this is not always the case, our fifo only have 16bytes and it can easily be filled up if there is too much logging, it will stuck at this while(1) loop waiting for fifo to be not full. This will cause performance issue.

With this PR, rtl8730e_log_up_txready will check if tx fifo is not full first, so in uart_xmitchars, it will only call loguart_putchar to send data when fifo is not full. This prevents it from entering the while(1) loop in loguart_putchar waiting fifo to be not full.. if fifo full, wait for next irq to do the tx.

but for up_putc, it is called by lldbg and other low level debugging apis which does not involves irq, so it requires the busy-wait to ensure that fifo has available room to put data.

Hope this can clarify your doubts.